doc/reading/tma-modeling-in-depth.md

```diff
@@ -0,0 +1,153 @@
+            gmem_idx1 = x[1] + i1 * er[1]
+            gmem_idx = gmem_idx0 * gr[0] + gmem_idx1 * gr[1]
+            if op == "load":
+                if gmem_idx0 < gs[0] and gmem_idx1 < gs[1]:
+                    sa[smem_idx] = ga[gmem_idx]
+                else:
+                    sa[smem_idx] = 0
+            else:
+                if gmem_idx0 < gs[0] and gmem_idx1 < gs[1]:
+                    ga[gmem_idx] = sa[smem_idx]
+```
+
+## Predication and correctness
+
+We all know that TMA has built-in boundary check and automatic zero filling for out-of-boundary items.
+But what does this mean? Does it mean that we don't need to predicate a TMA expression at all?
+Does it mean that we don't need to initialize the output buffer of TMA?
+This section discusses these questions.
+
+First to note is, what we should do depend on how we will use the output of TMA.
+The theory for this is called [correctness model](divisibility-of-split.md#allocation-and-correctness-model).
+Specifically, we are interested in:
+
+- For an unpredicated TMA expression, if we do not initialize the buffer, what correctness model can we achieve?
+- In the case where strong correctness is needed, but we have only achieved weak correctness,
+  is there any way to upgrade to strong correctness by doing something?
+
+As we can see from CodeBlock 1, TMA has builtin predicates checking that the indices of all partitioned IterDomains are in bound.
+That is, TMA will never do out-of-boundary access on global memory even if the indices of
+some IterDomains may be out of boundary.
+Therefore, we have:
+
+**Theorem 1:** TMA provides weak correctness.
+
+Having weak correctness is great in the sense that,
+if weak correctness is sufficient for us,
+we neither need to predicate nor need to initialize the output of TMA to achieve correct functionality.
+
+The following example Figure 2 shows a schedule of the consumer of a TMA load.
+From this figure, we can see that, if no additional predicate is used to guard the TMA load expression,
+some holes in allocation domain will be filled with in boundary data,
+and some holes will be filled with out of boundary zeros:
+
+![Figure 2: Holes in allocation domain filled with in boundary data](tma-modeling-in-depth/weak-correctness-holes-nonzero.svg)
+
+A common use case for TMA is to load data for tensor core.
+Because tensor core has unpredicated reduction, strong correctness is required.
+Strong correctness means that, when an IterDomain expression create holes (indivisible split, resize),
+the holes must be filled with a certain value.
+
+Understanding strong correctness of TMA requires the following definition:
+
+**Definition 1 (TMA-protected IterDomain):** An IterDomain is TMA-protected if it satisfies any of the following conditions:
+
+1. It is a partitioned IterDomain.
+2. It is the outer output of a split of a TMA-protected IterDomain.
+3. It is the output of a merge whose outer input is a TMA-protected IterDomain.
```

Comment:
Could you give an example where the inner input is not protected but the output is protected due to the protected outer input?

Reply:
See my latest commit: https://github.com/NVIDIA/Fuser/pull/2041/commits/036a495f10836e9e984f7c9ad43c730e57a28d79

Reply:
I think the analysis you are referring to is exactly the analysis described here. In this language, `I1` is TMA-protected, but `I2` is not, so `I2` needs to be predicated if there is an indivisible on it.

Being `TMA-protected` has nothing to do with whether it can go out of boundary or not. They are orthogonal concepts. Being `TMA-protected` means that, if it can go out of boundary, then it will be protected by the hardware.

|                        | TMA-protected  | not TMA-protected    |
| ---------------------- | -------------- | -------------------- |
| **can go out of bound**    | no predication | should be predicated |
| **cannot go out of bound** | no predication | no predication       |

Reply:
Added a new commit trying to explain why we want this definition: https://github.com/NVIDIA/Fuser/pull/2041/commits/0959cf0dd90e8fe0b9350df4614b3b7f44c961ea

Reply:
Thanks. That makes sense.